morango/utils/syncsession.py

```diff
@@ -0,0 +1,111 @@
+        """
+        where_statements = []
+        # create condition for all incoming FSICs where instance_ids are equal, but internal counters are higher then FSICs counters
+        for instance, counter in iteritems(fsics):
+            where_statements += ['(last_saved_instance == {0} AND last_saved_counter > {1})'.format(instance, counter)]
+        condition = _join_with_logical_operator(where_statements, 'OR')
+        # execute raw sql to take all records that match condition, to be put into buffer for transfer
+        with connection.cursor() as cursor:
+            queue_buffer = '''INSERT INTO {outgoing}
+                            (serialized, deleted, last_saved_instance, last_saved_counter,
+                             model_name, profile, partitions, transfer_session_id, incoming_buffer)
+                            SELECT serialized, deleted, last_saved_instance, last_saved_counter,
+                            model_name, profile, partitions, {transfer_session_id}, {incoming_buffer}
+                            FROM {store}
+                            [WHERE {condition}]'''.format(outgoing=DataTransferBuffer._meta.db_table,
+                                                          transfer_session_id=self.transfer_session_id,
+                                                          condition=condition,
+                                                          store=Store._meta.db_table,
+                                                          incoming_buffer=False)
+            cursor.execute(queue_buffer)
+            # take all record max counters that are foreign keyed onto store models, which were queued into the buffer
+            queue_rmc_buffer = '''INSERT INTO {outgoing_rmc}
+                                (instance_id, counter, incoming_buffer, transfer_session_id, store_model_id)
+                                SELECT instance_id, counter, {incoming_buffer}, {transfer_session_id}, store_model_id
+                                FROM {record_max_counter} as rmc
+                                WHERE EXISTS (SELECT *
+                                              FROM {outgoing_buffer} as buffer
+                                              WHERE buffer.store_model_id == rmc.store_model_id)
+                                '''.format(outgoing_rmc=RecordMaxCounterBuffer._meta.db_table,
+                                           incoming_buffer=False,
+                                           transfer_session_id=self.transfer_session_id,
+                                           record_max_counter=RecordMaxCounter._meta.db_table,
+                                           outgoing_buffer=DataTransferBuffer._meta.db_table)
+            cursor.execute(queue_rmc_buffer)
+
+    @transaction.atomic
+    def _integrate_into_store(self, session):
+        """
+        Takes data from the buffers and merges into the store and record max counters.
+        """
```

Comment:
We'll want to do all the "fast forwards" as a single SQL query block transfer (that can just overwrite the store entries, since it's a fast forward), and then another query to get all the ones that are merge conflicts -- and those we *will* need to loop over individually.